### PR TITLE
BRITECHSRM-152: Trello sync/fetch members

### PR DESCRIPTION
**Pull request by jebbinSCL**

### CODE REVIEW

File: TrelloSync/TrelloSync/App.config [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/App.config]: Should this App.config or app.config? If you create a new solution in Visual Studio on Windows, it's called App.config, but it's app.config in our microservice projects.

File: TrelloSync/TrelloSync/Helpers.fs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿module Helpers
+    open System.Configuration
+    type TrelloCredentials = 
+        { Key : string
+          Token : string } 
+
+    let credConfigTemplate = """<?xml version="1.0" encoding="utf-8" ?>
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Helpers.fs]: Why is this necessary? The app.config is part of the repo.

File: TrelloSync/TrelloSync/Helpers.fs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+                            <appSettings>
+                                <add key ="TrelloKey" value="{UserTrelloKey}" />
+                                <add key ="TrelloToken" value="{UserTrelloToken}" />
+                            </appSettings>"""
+
+    let missingCredentailsMessage = "TrelloKey or TrelloToken is unset. Create a TrelloCreds.config file next to the App.config with the following xml (Filling in the key and token value fields):" + credConfigTemplate
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Helpers.fs]: "credentials"

File: TrelloSync/TrelloSync/AssemblyInfo.fs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+// Setting ComVisible to false makes the types in this assembly not visible 
+// to COM components.  If you need to access a type in this assembly from 
+// COM, set the ComVisible attribute to true on that type.
+[<assembly: ComVisible(false)>]
+
+// The following GUID is for the ID of the typelib if this project is exposed to COM
+[<assembly: Guid("621dbe5c-de53-4bb7-9983-5359942556b8")>]
+
+// Version information for an assembly consists of the following four values:
```
Comment (thullSL) [inline TrelloSync/TrelloSync/AssemblyInfo.fs]: Are all these comment necessary?

File: TrelloSync/TrelloSync/App.config (hunk 1 above)
Comment (jebbinSCL) [inline TrelloSync/TrelloSync/App.config]: At this point I went with app.config for consistency, but it doesn't make a difference as long as you specify it correctly in the project file. We can have an item to change it in all repos later if we want.

File: TrelloSync/TrelloSync/Helpers.fs (hunk 3 above)
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Helpers.fs]: Do we need TrelloCreds.config? We have UNSET in app.config in this PR. The user of the app just needs to populate this.

File: TrelloSync/TrelloSync/Helpers.fs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+﻿module Helpers
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Helpers.fs]: This module is about credentials? Helpers doesn't allow us to read code that uses this module very easily.

File: TrelloSync/TrelloSync/Members.fs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        | 1  -> 
+            split.[0] + "@scottlogic.co.uk"
+        | x when x > 1 -> 
+            let firstNameFirstLetter = split.[0].Chars(0)
+            let lastName = (Array.last split).ToLowerInvariant()
+            sprintf "%c%s@scottlogic.co.uk" firstNameFirstLetter lastName
+        | _ -> "missingEmail@scottlogic.co.uk" 
+
+    let createImageUrl (avatarHash : string) =
+        sprintf "https://trello-avatars.s3.amazonaws.com/%s/50.png" <| avatarHash
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Members.fs]: Why do we need a reverse pipe?

File: TrelloSync/TrelloSync/Helpers.fs (hunk 3 above)
Comment (jebbinSCL) [inline TrelloSync/TrelloSync/Helpers.fs]: The trelloCreds.config is an ignored file, hopefully making it harder to checkin credentials. If its in the app, then we have to make sure to take out the credentials if we make any change to the app.config?

File: TrelloSync/TrelloSync/Program.fs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+﻿open System.IO
+open Newtonsoft.Json
+
+[<EntryPoint>]
+let main argv = 
+    let trelloCred = Helpers.getTrelloCredentials()
+    let trelloData = Members.getAllMembersAsync trelloCred |> Async.RunSynchronously
+    let result = JsonConvert.SerializeObject(trelloData, Formatting.Indented)
+    File.WriteAllText(@"trello-import.json",result)
+    printfn "%A" argv
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Program.fs]: Why do we need this line?

File: TrelloSync/TrelloSync/Helpers.fs (hunk 2 above)
Comment (jebbinSCL) [inline TrelloSync/TrelloSync/Helpers.fs]: To make it clear that only this section is needed for the entire TrelloCred.config file, and so that anyone can just copy paste.

File: TrelloSync/TrelloSync/AssemblyInfo.fs (hunk 4 above)
Comment (jebbinSCL) [inline TrelloSync/TrelloSync/AssemblyInfo.fs]: They are added automatically. I'll leave them since its useful information.

File: TrelloSync/TrelloSync/Helpers.fs (hunk 3 above)
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Helpers.fs]: That's fine. We do that for credentials.yml in the concourse configuration.

File: TrelloSync/TrelloSync/Members.fs [hunk 8]
```diff
@@ -0,0 +1,88 @@
+﻿module Members
+    open System
+    open System.Net
+    open Newtonsoft.Json 
+    open Helpers
+
+    type BasicMember =
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Members.fs]: What's the difference between a ```BasicMember``` and a ```TrelloMember```?

File: TrelloSync/TrelloSync/Members.fs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+          AvatarHash : string }
+    
+    type TrelloMember = 
+        { Id : string
+          Name : string
+          Email : string
+          UserName : string
+          ImageUrl : string }
+
+    type MembersMeta =
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Members.fs]: Is this metadata?

File: TrelloSync/TrelloSync/Members.fs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+          ImageUrl : string }
+
+    type MembersMeta = 
+        { Members : TrelloMember []
+          IgnoredMembers : TrelloMember []
+          DefaultMember : TrelloMember }
+
+    type MemberId = { Id: string}
+
+    let parseToEmail (fullName : string) =
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Members.fs]: It would probably be worth a comment to explain the rules here.

File: TrelloSync/TrelloSync/Members.fs (hunk 8 above)
Comment (jebbinSCL) [inline TrelloSync/TrelloSync/Members.fs]: Basic member is the record used in the deserialization, while trelloMember contains additional fields that we add after parsing the basic member data.

File: TrelloSync/TrelloSync/Members.fs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    let parseToEmail (fullName : string) = 
+        let split = fullName.Split()
+        match split.Length with 
+        | 1  -> 
+            split.[0] + "@scottlogic.co.uk"
+        | x when x > 1 -> 
+            let firstNameFirstLetter = split.[0].Chars(0)
+            let lastName = (Array.last split).ToLowerInvariant()
+            sprintf "%c%s@scottlogic.co.uk" firstNameFirstLetter lastName
+        | _ -> "missingEmail@scottlogic.co.uk"
```
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Members.fs]: Was this agreed upon?

File: TrelloSync/TrelloSync/Members.fs (hunk 8 above)
Comment (jebbinSCL) [inline TrelloSync/TrelloSync/Members.fs]: Happy to rename if you can think of a clearer name.

File: TrelloSync/TrelloSync/Members.fs (hunk 9 above)
Comment (jebbinSCL) [inline TrelloSync/TrelloSync/Members.fs]: Not really, but I was having trouble thinking of a clearer name, other than "Wrapper" of some sort

File: TrelloSync/TrelloSync/Members.fs (hunk 11 above)
Comment (jebbinSCL) [inline TrelloSync/TrelloSync/Members.fs]: The missingEmail? It's something that should never bee seen since we are creating the name from the email anyway. Actually, its basically an impossible state, so I'll change this to a exception if you agree?

File: TrelloSync/TrelloSync/Members.fs (hunk 9 above)
Comment (csscottlogic) [inline TrelloSync/TrelloSync/Members.fs]: It's a categorisation

File: TrelloSync/TrelloSync/App.config (hunk 1 above)
Comment (dwybourn) [inline TrelloSync/TrelloSync/App.config]: Do you mean you went with App.config (not app.config)?

Comment (jebbinSCL) [inline TrelloSync/TrelloSync/App.config]: Ah. I'll leave it as is for now, then add an item to deal with this for all projects

**Comment by dwybourn** [comment]
:shipit:
